Develop

karvalolka opened the pull request.

karvalolka committed:
+ авторизация;
+ bootstrap;
+ БД;
+ миграции;

karvalolka committed:
исправлен нейминг колонок в БД (ru->eng);
в tab char + поле xp;
изправлена ошибка в формате нейминга поля в БД (snake_case);
+ привязка (char+inventary, race+fraction, (protection_areas+armor/weapon);
+ модели к миграциям
переименнованы модель и миграция (class->grade)
+ модель и миграция (protection_areas)

kovdmit commented on database/migrations/2024_09_04_160226_create_chars_table.php:
```diff
@@ -0,0 +1,34 @@
+        Schema::create('chars', function (Blueprint $table) {
+            $table->id();
+            $table->string('nickname');
+            $table->unsignedTinyInteger('level');
```
В ТЗ не было поля уровень. Было только опыт. Ну ладно, пусть уже остаётся

karvalolka replied:
я изначально подумал, что ты написал опыт это значит уровень

kovdmit replied:
Лвл по сути можно динамически вычислять на основе опыта. Например если опыт, скажем = 13579, то 13 лвл. Чё то такое

kovdmit commented on database/migrations/2024_09_04_160226_create_chars_table.php:
```diff
@@ -0,0 +1,34 @@
+            $table->unsignedMediumInteger('mp');
+            $table->unsignedSmallInteger('attack');
+            $table->unsignedSmallInteger('def');
+            $table->unsignedInteger('xp');
```
Не критично: Windows XP?) Опыт обозначают либо experience либо exp

kovdmit commented on database/migrations/2024_09_04_160356_create_inventories_table.php:
```diff
@@ -0,0 +1,30 @@
+        Schema::create('inventories', function (Blueprint $table) {
+            $table->id();
+            $table->unsignedInteger('gold');
+            $table->text('loot');
```
Для чего это поле?

karvalolka replied:
для золота

kovdmit commented on database/migrations/2024_09_04_160356_create_inventories_table.php:
```diff
@@ -0,0 +1,30 @@
+            $table->foreignId('char_id')->unique()->constrained()->onDelete('cascade');
+            $table->timestamps();
+        });
+    }
```
А как же 5 ячеек для инвентаря?

kovdmit commented on database/migrations/2024_09_04_173925_create_armors_table.php:
```diff
@@ -0,0 +1,30 @@
+        Schema::create('armors', function (Blueprint $table) {
+            $table->id();
+            $table->string('name');
+            $table->unsignedSmallInteger('def');
```
Неочевидное наименование. Используй value например

karvalolka replied:
где

kovdmit replied:
def

kovdmit commented on database/migrations/2024_09_04_173952_create_weapons_table.php:
```diff
@@ -0,0 +1,30 @@
+        Schema::create('weapons', function (Blueprint $table) {
+            $table->id();
+            $table->string('name');
+            $table->unsignedSmallInteger('attack');
```
Неочевидное наименование

karvalolka replied:
где

kovdmit replied:
Attack

kovdmit commented on database/migrations/2024_09_04_173952_create_weapons_table.php:
```diff
@@ -0,0 +1,30 @@
+            $table->id();
+            $table->string('name');
+            $table->unsignedSmallInteger('attack');
+            $table->foreignId('protection_area_id')->constrained()->onDelete('cascade');
```
Как область защиты связана с оружием?

kovdmit commented on app/Models/Char.php:
```diff
@@ -0,0 +1,19 @@
+    {
+        return $this->hasOne(Inventory::class);
+    }
+}
```
Нет метода для получения класса

kovdmit commented on app/Models/Inventory.php:
```diff
@@ -0,0 +1,19 @@
+    {
+        return $this->belongsTo(Char::class);
+    }
+}
```
Там шо то было про 5 ячеек

kovdmit commented on app/Models/Weapon.php:
```diff
@@ -0,0 +1,19 @@
+    protected $table = 'weapons';
+    protected $guarded = false;
+
+    public function protectionArea()
```
Никак не могу понять, как оружие связано с областью защиты))

karvalolka committed:
изменено отношение Char и Inventory;
изменина миграция создания инвентаря;
xp->exp

karvalolka committed:
отвязал оружие отпоказателя защиты;
+ миграция и модель показателя атаки;
+ привязка с оружием;
'attack'->'power', 'def'->'power';
+ метода для получения grade в Char;
+ метод и модель для динамического создания слотов в инвентаре;
изменение логики в добавлении/удалении слотов;
+ динамический уровень от exp;

kovdmit commented on app/Models/Char.php:
```diff
@@ -0,0 +1,29 @@
+
+    public function inventory()
+    {
+        return $this->hasMany(Inventory::class);
```
У персонажа может быть много инвентарей?

gotyefrid replied:
))инвентарь твой и инвентарь квестов - как в пв)))

kovdmit commented on app/Models/Inventory.php:
```diff
@@ -0,0 +1,50 @@
+    }
+    public function addItem($item)
+    {
+        if ($this->slots()->count() < 5) {
```
Плохая идея использовать "магические числа".

kovdmit commented on app/Models/Inventory.php:
```diff
@@ -0,0 +1,50 @@
+    public function addItem($item)
+    {
+        if ($this->slots()->count() < 5) {
+            $this->slots()->create(['item' => $item]);
```
Я конечно не ларавельщик, но Шо то я не думаю, что у тебя получится так добавить предмет в инвентарь. 

Метод slots вернёт тебе коллекцию имеющихся слотов. И у этой коллекции вряд ли есть метод create.

Если я прав, Можешь оставить реализацию на дальнейшую разработку за рамками данного PR. Но тогда для этого удали этот метод.

Если не прав, то сори, я путхонист, у меня лапки

karvalolka replied:
"Друг" говорит, что должно работать, предлагаю уже на практике в будущем проверить

kovdmit replied:
Принято

kovdmit commented on app/Models/Inventory.php:
```diff
@@ -0,0 +1,50 @@
+        return "Некорректный номер слота!";
+    }
+
+    public function updateGold($amount)
```
Из названия метода непонятно, что он будет именно складывать старое значение с новым, а не просто перезаписывать старое

kovdmit commented on database/migrations/2024_09_04_160226_create_chars_table.php:
```diff
@@ -0,0 +1,33 @@
+            $table->string('nickname');
+            $table->unsignedMediumInteger('hp');
+            $table->unsignedMediumInteger('mp');
+            $table->unsignedSmallInteger('attack');
```
Допиши плиз attack_power,  раз уж пошла такая пьянка

kovdmit commented on database/migrations/2024_09_04_160226_create_chars_table.php:
```diff
@@ -0,0 +1,33 @@
+            $table->unsignedMediumInteger('hp');
+            $table->unsignedMediumInteger('mp');
+            $table->unsignedSmallInteger('attack');
+            $table->unsignedSmallInteger('def');
```
Тоже

kovdmit commented on database/migrations/2024_09_07_170013_create_inventory_slots_table.php:
```diff
@@ -0,0 +1,30 @@
+    {
+        Schema::create('inventory_slots', function (Blueprint $table) {
+            $table->id();
+            $table->string('item')->nullable();
```
Вот непонятно почему строка. Тут же должна быть ссылка на шмот.

karvalolka committed:
изменил связть часа с инвентарём;
- магические числа;
updateGold->changeGold, attack->attack_power, def->def_power;

kovdmit commented on app/Models/Inventory.php:
```diff
@@ -0,0 +1,52 @@
+        return "Некорректный номер слота!";
+    }
+
+    public function changeGold($amount)
```
Ни "update", ни "change" не указывает на то, что передаваемое значение будет складываться со старым. 

Предлагаю использовать что-то вроде глагола "увеличить" или "сложить".

karvalolka committed:
+ миграция и модель item;
ссылка на item в inventory_slots;
изменена формула вычисления голды;
+ test route

karvalolka committed:
swap bootstrap

karvalolka committed:
оптимизация шаблона;
+ стили для шаблона;
+ админка;
отредактировал админку;

karvalolka committed:
+ route index и create;
+ sidebar (для левой панели у админки);
+ @include;
+ Controllers create и index
